message.go

```diff
@@ -480,9 +480,43 @@ type MultiSegmentArena [][]byte
 // they are full.  b MAY be nil.  Callers MAY use b to populate the
 // buffer for reading or to reserve memory of a specific size.
 func MultiSegment(b [][]byte) *MultiSegmentArena {
+	if b == nil {
+		return multiSegmentPool.Get().(*MultiSegmentArena)
+	}
+	return multiSegment(b)
+}
+
+// Return this arena to an internal sync.Pool of arenas that can be
+// re-used. Any time MultiSegment(nil) is called, arenas from this
+// pool will be used if available, which can help reduce memory
+// allocations. Calling Release however is optional; if not done
+// the garbage collector will release the memory per usual.
+func (msa *MultiSegmentArena) Release() {
+	for i, v := range *msa {
+		// Clear the memory, so there's no junk in here for the next use:
+		for j, _ := range v {
+			v[j] = 0
+		}
+
+		// Truncate the segment, since we use the length as the marker for
+		// what's allocated:
+		(*msa)[i] = v[:0]
+	}
+	(*msa) = (*msa)[:0] // Hide the segments
+	multiSegmentPool.Put(msa)
```

Comment:
A frequent issue with `sync.Pool` is that large objects may be kept alive despite not being used to their fullest capacity, resulting in high memory usage.  The following mitigation strategies are common:

1. drop objects larger than some arbitrary size,
2. maintain separate pools for objects of similar size, and;
3. try to estimate the "drop" threshold dynamically ([example](https://pkg.go.dev/github.com/libp2p/go-buffer-pool))

We may wish to kick this particular can down the road, but we should probably open an issue to track it.  Perhaps in the interim, we might use strategy 1 with a generous cutoff?

Reply:
Clarification:  strategy 2 has the effect of increasing the probability of large objects being GCed.  I've seen it mixed and matched with strategies 1 & 3.

Reply:
In the interim, I don't think it's likely to be a problem for RPC workloads, since there's going to be a cap on message sizes anyway. But yeah, we could open an issue to track this stuff.

Reply:
Addendum:  we might also consider applying these strategies on a per-segment basis.

Reply:
Ah yeah, that's a good point.  Great!  👍 

Resolving.

Reply:
Opened #351 for this.